File: src/main/java/core/procedures/sets/Intersection.java
package core.procedures.sets;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.HashSet;
import java.util.Set;

public final class Intersection extends AFn {

  public Intersection() {
    super(new FnArgsBuilder().min(1).mandatory(new Class[]{Set.class}).rest(Set.class).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "intersection";
  }

  @Override
  public Set<Object> apply(Object... args) {
    if (args.length == 1) {
      return (Set<Object>) args[0];
    }
    Set result = new HashSet((Set)args[0]);
    for (int i = 1, argsLength = args.length; i < argsLength; i++) {
      result.retainAll((Set)args[i]);
    }
    return result;
  }
}

File: src/main/java/core/procedures/sets/Intersection.kt
package core.procedures.sets

import core.procedures.AFn
import core.scm.MutableSet

class Intersection : AFn<Any?, Set<*>>(name = "intersection", isPure = true, minArgs = 1,
                                       mandatoryArgsTypes = arrayOf(Set::class.java), restArgsType = Set::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when {
        args.size == 1 -> args[0]!! as Set<*>
        else -> MutableSet(args[0]!! as Set<*>).apply {
            for (i in (1 until args.size)) {
                retainAll(args[i]!! as Set<*>)
            }
        }
    }
}
